File: frontend/lib/timelineSimulator.ts
import type { RiskLevel, TimelineAnalysis } from "@/types/analysis";

/**
 * Timeline risk simulator (pure functions — easy to unit test and to
 * mirror in backend/services/workflow/timeline_simulator.py).
 */

export type TimelineInputs = {
  organizerResponseDays: number;
  advisorResponseDays: number;
  studentDecisionDays: number;
};

export const DEFAULT_TIMELINE_INPUTS: TimelineInputs = {
  organizerResponseDays: 2,
  advisorResponseDays: 2,
  studentDecisionDays: 1,
};

export type TimelineResult = {
  daysUntilDeadline: number | null;
  daysNeeded: number;
  riskLevel: RiskLevel;
  recommendation: string;
  /** Ordered verification steps with their day estimates. */
  steps: { label: string; days: number; owner: string }[];
  /** Latest calendar date the student should start asking questions. */
  latestAskDate: string | null;
  /** ISO date the verification chain would finish if started today. */
  verificationReadyDate: string;
};

/** Days from today (local midnight) until an ISO date. Null if unparseable. */
export function daysUntil(dateStr: string | null, today: Date = new Date()): number | null {
  if (!dateStr) return null;
  const target = new Date(dateStr);
  if (isNaN(target.getTime())) return null;
  const a = new Date(today.getFullYear(), today.getMonth(), today.getDate());
  const b = new Date(target.getFullYear(), target.getMonth(), target.getDate());
  return Math.round((b.getTime() - a.getTime()) / 86_400_000);
}

function addDays(date: Date, days: number): Date {
  const d = new Date(date);
  d.setDate(d.getDate() + days);
  return d;
}

function formatDate(d: Date): string {
  return d.toLocaleDateString("en-US", { month: "short", day: "numeric" });
}

/**
 * Core risk rules:
 *  - no deadline           -> low (but still recommend verifying)
 *  - deadline passed/today -> high
 *  - need > have           -> high
 *  - need == have, or within 2 days of have -> medium_high
 *  - within 5 days of have -> moderate
 *  - otherwise             -> low
 */
export function simulateTimeline(
  deadlineOrStartDate: string | null,
  inputs: TimelineInputs = DEFAULT_TIMELINE_INPUTS,
  today: Date = new Date()
): TimelineResult {
  const steps = [
    { label: "Organizer response", days: inputs.organizerResponseDays, owner: "organizer" },
    { label: "Advisor/DSO clarification", days: inputs.advisorResponseDays, owner: "advisor" },
    { label: "Student decision", days: inputs.studentDecisionDays, owner: "student" },
  ];
  const daysNeeded = steps.reduce((sum, s) => sum + s.days, 0);
  const daysAvailable = daysUntil(deadlineOrStartDate, today);
  const verificationReadyDate = formatDate(addDays(today, daysNeeded));

  if (daysAvailable === null) {
    return {
      daysUntilDeadline: null,
      daysNeeded,
      riskLevel: "low",
      recommendation:
        "No deadline detected. You still should verify eligibility before applying, but timing is not the main risk.",
      steps,
      latestAskDate: null,
      verificationReadyDate,
    };
  }

  let riskLevel: RiskLevel;
  let recommendation: string;
  const slack = daysAvailable - daysNeeded;
  const latestAskDate =
    slack >= 0 ? formatDate(addDays(today, slack)) : formatDate(today);

  if (daysAvailable <= 0) {
    riskLevel = "high";
    recommendation =
      "The deadline or start date has already arrived. Contact the organizer immediately to ask if late verification or applications are possible.";
  } else if (slack < 0) {
    riskLevel = "high";
    recommendation = `Verification needs about ${daysNeeded} days but only ${daysAvailable} remain. Ask the organizer right now and flag the deadline in your message.`;
  } else if (slack <= 2) {
    riskLevel = "medium_high";
    recommendation = `You have ${daysAvailable} days and verification may take ${daysNeeded}. Ask the organizer today — verification may take almost the entire remaining time.`;
  } else if (slack <= 5) {
    riskLevel = "moderate";
    recommendation = `You have ${daysAvailable} days and verification takes about ${daysNeeded}. Start verification by ${latestAskDate} to stay safe.`;
  } else {
    riskLevel = "low";
    recommendation = `You have ${daysAvailable} days — enough time to verify (about ${daysNeeded} days). Still, ask early: ${latestAskDate} is your latest safe start.`;
  }

  return {
    daysUntilDeadline: daysAvailable,
    daysNeeded,
    riskLevel,
    recommendation,
    steps,
    latestAskDate,
    verificationReadyDate,
  };
}

/**
 * Identify the bottleneck: the step with the largest day estimate
 * (first one wins ties — earlier steps block everything downstream).
 */
export function findBottleneck(result: TimelineResult) {
  return result.steps.reduce((max, s) => (s.days > max.days ? s : max), result.steps[0]);
}

/** Convert a simulation into the shared TimelineAnalysis shape. */
export function toTimelineAnalysis(
  deadlineOrStartDate: string | null,
  result: TimelineResult
): TimelineAnalysis {
  return {
    deadline_or_start_date: deadlineOrStartDate,
    days_until_deadline: result.daysUntilDeadline,
    estimated_verification_days: result.daysNeeded,
    risk_level: result.riskLevel,
    critical_path: result.steps.map((s) => s.label),
    recommendation: result.recommendation,
  };
}

File: backend/services/workflow/timeline_simulator.py
# Timeline risk simulator.
# Mirrors frontend/lib/timelineSimulator.ts. Returns the shared-schema
# "timeline" object.

from __future__ import annotations

from datetime import date, datetime, timedelta

DEFAULT_INPUTS = {
    "organizer_response_days": 2,
    "advisor_response_days": 2,
    "student_decision_days": 1,
}

STEPS = [
    ("Organizer response", "organizer_response_days"),
    ("Advisor/DSO clarification", "advisor_response_days"),
    ("Student decision", "student_decision_days"),
]


def days_until(date_str: str | None, today: date | None = None) -> int | None:
    """Days from today until an ISO-ish date string. None if unparseable."""
    if not date_str:
        return None
    today = today or date.today()
    for fmt in ("%Y-%m-%d", "%m/%d/%Y", "%B %d, %Y", "%b %d, %Y"):
        try:
            target = datetime.strptime(date_str.strip(), fmt).date()
            return (target - today).days
        except ValueError:
            continue
    return None


def simulate_timeline(
    deadline_or_start_date: str | None,
    inputs: dict[str, int] | None = None,
    today: date | None = None,
) -> dict:
    """Compute timeline risk in the shared schema.

    Risk rules (keep in sync with the frontend):
      - no deadline           -> low
      - deadline passed/today -> high
      - need > have           -> high
      - slack <= 2 days       -> medium_high
      - slack <= 5 days       -> moderate
      - otherwise             -> low
    """
    inputs = {**DEFAULT_INPUTS, **(inputs or {})}
    today = today or date.today()

    days_needed = sum(inputs[key] for _, key in STEPS)
    days_available = days_until(deadline_or_start_date, today)
    critical_path = [label for label, _ in STEPS]

    base = {
        "deadline_or_start_date": deadline_or_start_date,
        "days_until_deadline": days_available,
        "estimated_verification_days": days_needed,
        "critical_path": critical_path,
    }

    if days_available is None:
        return {
            **base,
            "risk_level": "low",
            "recommendation": (
                "No deadline detected. You still should verify eligibility before "
                "applying, but timing is not the main risk."
            ),
        }

    slack = days_available - days_needed
    latest_ask = (today + timedelta(days=max(slack, 0))).strftime("%b %d")

    if days_available <= 0:
        risk, rec = "high", (
            "The deadline or start date has already arrived. Contact the organizer "
            "immediately to ask if late verification or applications are possible."
        )
    elif slack < 0:
        risk, rec = "high", (
            f"Verification needs about {days_needed} days but only {days_available} "
            "remain. Ask the organizer right now and flag the deadline in your message."
        )
    elif slack <= 2:
        risk, rec = "medium_high", (
            f"You have {days_available} days and verification may take {days_needed}. "
            "Ask the organizer today — verification may take almost the entire "
            "remaining time."
        )
    elif slack <= 5:
        risk, rec = "moderate", (
            f"You have {days_available} days and verification takes about "
            f"{days_needed}. Start verification by {latest_ask} to stay safe."
        )
    else:
        risk, rec = "low", (
            f"You have {days_available} days — enough time to verify (about "
            f"{days_needed} days). Still, ask early: {latest_ask} is your latest "
            "safe start."
        )

    return {**base, "risk_level": risk, "recommendation": rec}
